### PR TITLE
Issue #6974 - improvements & fixes to ByteBufferPool implementations (#7017)

### PR DESCRIPTION
**Pull request by lachlan-roberts**
## Merge from `9.4.x` to `10.0.x` of PR #7017.

- WebSocket should user server ByteBufferPool if possible
- fix various bugs ByteBufferPool implementations
- add heuristic for maxHeapMemory and maxDirectMemory
- Add dump for ByteBufferPools
- add LogArrayByteBufferPool that does exponential scaling of bucket size.
- ByteBufferPools should default to use maxMemory heuristic
- Add module jetty-bytebufferpool-logarithmic

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>
Co-authored-by: Simone Bordet <simone.bordet@gmail.com>

### CODE REVIEW

**Comment by gregw** [comment]
@lachlan-roberts not building?

**Comment by gregw** [comment]
Also, is this a direct port from 9.4, or is anything different?

**Comment by lachlan-roberts** [comment]
@gregw this is a direct port from 9.4 with no additional changes.

There is a test failing: `org.eclipse.jetty.http2.server.HTTP2ServerTest`. 
I will investigate soon.

**Comment by gregw** [comment]
@lachlan-roberts still not building and last 2 commits look kind of fundamental changes???

**Review by gregw — CHANGES_REQUESTED** [review]
maybe make it a draft until it passes?

**Comment by lachlan-roberts** [comment]
@gregw this is not a draft and is ready for review, I have removed WebSocketBufferPoolTest as it is flakey.

The main change from the 9.4 PR is that I have made bucket 0 for pooling buckets with capacity 0.
And now the `int bucketFor(int capacity)` in `ArrayByteBufferPool` will return the index of the bucket and not the bucket "number". So each bucket will contain buffers with capacities of `bucketIndex * capacityFactor`.

**Comment by gregw** [comment]
@lachlan-roberts still failing

File: jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java [hunk 1]
```diff
@@ -104,13 +104,13 @@ public ArrayByteBufferPool(int minCapacity, int factor, int maxCapacity, int max
         _minCapacity = minCapacity;
 
         // Initialize all buckets in constructor and never modify the array again.
-        int length = bucketFor(maxCapacity);
+        int length = bucketFor(maxCapacity) + 1;
```
Comment (gregw) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:107]: what is the extra allocated bucket used for?

Comment (lachlan-roberts) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:107]: @gregw `bucketFor(capacity)` gives you the index of the bucket of that capacity. So the size of the array is `bucketFor(maxCapacity) + 1`. The bucket at index 0 now contains the buffers of capacity 0 and I wasn't including this before.

Also the test failure was unrelated, the log said `Error cloning remote repo 'origin'`, so I have started a re-run of the tests.

File: jetty-server/src/main/config/etc/jetty-bytebufferpool-logarithmic.xml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0"?><!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure_9_3.dtd">
```
Comment (sbordet) [inline jetty-server/src/main/config/etc/jetty-bytebufferpool-logarithmic.xml]: Update the DTD to 10.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-server/src/main/config/etc/jetty-bytebufferpool-logarithmic.xml (hunk 2 above)
Comment (lachlan-roberts) [reply jetty-server/src/main/config/etc/jetty-bytebufferpool-logarithmic.xml]: done

**Comment by lachlan-roberts** [comment]
@gregw this is ready to be reviewed.

**Review by sbordet — APPROVED** [review]

**Review by gregw — APPROVED** [review]

**Pull request merged**